rpc/js/src/signaling-exchange.ts

```diff
@@ -0,0 +1,391 @@
+    uuid: string,
+    response: CallResponseInitStage
+  ): Promise<boolean> {
+    if (this.haveInitResponse) {
+      await this.sendError('got init stage more than once');
+      return false;
+    }
+    this.haveInitResponse = true;
+
+    this.callUuid = uuid;
+
+    const remoteSDP = new RTCSessionDescription(
+      JSON.parse(atob(response.sdp)) as RTCSessionDescriptionInit
+    );
+    if (this.clientChannel.isClosed()) {
+      await this.sendError('client channel is closed');
+      return false;
+    }
+    await this.pc.setRemoteDescription(remoteSDP);
+    this.awaitingRemoteDescription?.success(true);
+
+    if (this.dialOpts?.disableTrickleICE) {
+      this.exchangeDone = true;
+      await this.sendDone();
+      return false;
+    }
+
+    return true;
+  }
+
+  private async handleUpdateResponse(
+    uuid: string,
+    response: CallResponseUpdateStage
+  ): Promise<boolean> {
+    if (!this.haveInitResponse) {
+      await this.sendError('got update stage before init stage');
+      return false;
+    }
+
+    if (uuid !== this.callUuid) {
+      await this.sendError(`uuid mismatch; have=${uuid} want=${this.callUuid}`);
+      return false;
+    }
+    const cand = iceCandidateFromProto(response.candidate!);
+    if (cand.candidate !== undefined) {
+      console.debug(`received remote ICE ${cand.candidate}`);
+    }
+    try {
+      await this.pc.addIceCandidate(cand);
+    } catch (error) {
+      await this.sendError(JSON.stringify(error));
+      return false;
+    }
+
+    return true;
+  }
+
+  private async onLocalICECandidate(event: {
+    candidate: RTCIceCandidateInit | null;
+  }) {
+    await this.remoteDescriptionSet;
+    if (this.exchangeDone || this.pc.iceConnectionState === 'connected') {
+      return;
+    }
+
+    if (event.candidate === null) {
+      this.iceComplete = true;
+      await this.sendDone();
+      return;
+    }
+
+    if (!this.callUuid) {
+      throw new Error(callUUIDUnset);
+    }
+
+    if (event.candidate.candidate !== undefined) {
+      console.debug(`gathered local ICE ${event.candidate.candidate}`);
+    }
+    const iProto = iceCandidateToProto(event.candidate);
+    const callRequestUpdate = new CallUpdateRequest({
+      uuid: this.callUuid,
+      update: {
+        case: 'candidate',
+        value: iProto,
+      },
+    });
+    const callUpdateStart = new Date();
+    try {
+      await this.signalingClient.callUpdate(callRequestUpdate, this.callOpts);
+      this.numCallUpdates += 1;
+      const callUpdateEnd = new Date();
+      const callUpdateDuration =
+        callUpdateEnd.getTime() - callUpdateStart.getTime();
+      if (callUpdateDuration > this.maxCallUpdateDuration) {
+        this.maxCallUpdateDuration = callUpdateDuration;
+      }
+      this.totalCallUpdateDuration += callUpdateDuration;
+    } catch (error) {
+      if (
+        // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+        this.exchangeDone ||
```

Comment:
[Not a blocker, not the best time to resolve] Same comment here: it looks a lot like `this.exchangeDone` is already checked above, and I don't immediately see any method calls that would change `exchangeDone` between the two checks. Is something coming in externally and causing it to change?

Either way, I think if it's enough to confuse ESLint, it's enough to confuse a future reader like me.

Reply:
An await is used prior to exchange done bring checked which means based on how the event loop orders things, it may be set to true from another function. This is a case of unseen concurrency in JavaScript

Reply:
Without this extra check, you'll see a needs atomic check lint and the two resolutions to the lints are in conflict. I've seen this several times for concurrent code in JS and it's frustrating to work around.

Reply:
managed to simplify the code since there's no more factories here really